CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+
+        private void SetupCache()
+        {
+            double cleanupInterval = 100;
+            var cleanupTimer = new DefaultTimer(cleanupInterval);
+            _timeToLive = TimeSpan.FromSeconds(1);
+
+            _cache = CacheFactory.Build<DummyModel>(part => part
+                .WithProtoBufSerializer()
+                .WithSelfCleaningRedisConfiguration(_connectionMock.Object, cleanupTimer, _timeToLive,
+                    out string configurationKey)
+                .WithSelfCleaningRedisCacheHandle(configurationKey));
+
+            _onRemoveByHandleInvocations = new List<(CacheItemRemovedEventArgs, DateTime)>();
+            _cache.OnRemoveByHandle += (sender, args) => _onRemoveByHandleInvocations.Add((args, DateTime.Now));
+
+            _cache.CacheHandles.OfType<IStartable>().Single().Start();
+        }
+
+        #endregion
+
+        [TearDown]
+        public void TearDown()
+        {
+            _fauxDatabase.Clear();
+            _onRemoveByHandleInvocations.Clear();
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _cache.Dispose();
+        }
+
+        [TestCase(1, false)]
+        [TestCase(2, true)]
+        [TestCase(3, false)]
+        [TestCase(4, true)]
+        [TestCase(5, false)]
+        public void Put_AddSomeAmountOfKeys_OnRemoveByHandleCalledAfterTimeToLiveWithCorrectValueAndReason(
+            int amountOfKeys, bool delayBetweenInsertions)
+        {
+            // Arrange
+            IEnumerable<(string, DummyModel)> keys = Enumerable.Repeat("key", amountOfKeys)
+                .Select((key, i) => (key + i, new DummyModel {Property = "property" + i}));
+
+            TimeSpan delayBetweenInsertionsSpan = delayBetweenInsertions ? _timeToLive / 2 : TimeSpan.Zero;
+
+            // Act
+            var keysWithInsertionTimes = new List<(string, DummyModel, DateTime)>();
+
+            foreach ((string key, DummyModel value) in keys)
+            {
+                _cache[key] = value;
+                DateTime insertionTime = DateTime.Now;
+                keysWithInsertionTimes.Add((key, value, insertionTime));
+
+                Task.Delay(delayBetweenInsertionsSpan).Wait();
+            }
+
+            Task.Delay((int) (_timeToLive.TotalMilliseconds + TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD)).Wait();
+
+            // Assert
+            IDictionary<string, (CacheItemRemovedEventArgs Args, DateTime Time)> invocationsDictionary =
+                _onRemoveByHandleInvocations.ToDictionary(tuple => tuple.Args.Key, tuple => (tuple.Args, tuple.Time));
+
+            foreach ((string expectedKey, DummyModel expectedValue, DateTime insertionTime) in keysWithInsertionTimes)
+            {
+                CollectionAssert.Contains(invocationsDictionary.Keys, expectedKey);
+                OnRemoveByHandleAssertion(invocationsDictionary[expectedKey], expectedValue, insertionTime,
+                    _timeToLive);
+            }
+        }
```

Comment:
short comment: you have to verify (before the for) that keysWithInsertionTimes.count == invocationsDictionary.count

long comment:
this case will get true in this logic but should fail:
`keysWithInsertionTimes` has 2 values, which are included in `invocationsDictionary`, but `invocationsDictionary` has 1 more value - not a good and expected scenario and result

Reply:
changed